### PR TITLE
Readonly Futures and derivations

### PR DESCRIPTION
**Pull request by kourge**
This is ongoing work for https://github.com/UrbanDoor/mobx-failable/issues/2, which proposes _read-only futures_, which is the read-only variant of a `Future` (the common interface uniting `Failable` and `Loadable`), as well as _derived futures_, which are read-only futures whose state and value are twinned with the futures from which they derive. Rebases will occur frequently. Note that a large part of this is inspired by [`com.twitter.util.Future`](https://twitter.github.io/util/docs/com/twitter/util/Future.html).

### Happenings past, present, and future
Here is a combined checklist that is both a commit highlight reel as well as a to-do list:
- [x] 9f70a93, Introduce `ReadonlyFuture<T>`
- [x] d7aba5e, Reshape `Future.DeriveOptions`, which describes the option object that the slated `.derive(...)` method takes. The intent is that it should not be possible to say `.derive({})`.
- [x] e76adbe, Initial `DerivedFailable<T>` implementation. This is an internal class that consumers should not directly instantiate.
- [x] 58ecc23, Ensure that the mirroring of the underlying future happens consistently. For example, when a derived failable is constructed with only a success transform, it should still become a failure if its underlying future becomes a failure.
- [x] 1ba56fb, Lots of tests! Maybe not enough. Please comment if there are any missing edge cases.
- [x] 9a6f797, Implement and document `derive` and its shorthands `map` and `rescue`. See the link to the documentation for `com.twitter.util.Future` for comparison.
- [x] 74af47d, Write tests for the shorthand methods.
- [x] 137c8b7, Extend derivation support to `Loadable` as well. This involves restructuring the file system organization of it, just like `Failable`.
- [x] 315da0c, Add the three new methods into the interfaces `Future` and `ReadonlyFuture`.

### Examples

Transforming success values:

```ts
const asyncDates = new Failable<Date[]>();
const localeDates = asyncDates.map(dates => dates.map(d => d.toLocaleString()));
// something mutates `asyncDates`
// `localeDates` updates automatically
```

Transforming success to failure:

```ts
declare interface User { ... }

// Throws if `Server.User` is malformed
declare function verifyUser(in: Server.User): User;

const rawUserData = new Failable<Server.User>();
const user = rawUserData.map(verifyUser);
```

Please, all nits are welcome. Bring 'em in.

### CODE REVIEW

File: src/failable/derived.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  @observable protected data: To | Error | undefined = undefined;
+  @observable protected state: Future.State = State.pending;
+  protected transformation: IReactionDisposer;
+
+  constructor(
+    underlying: ReadonlyFuture<T>,
+    protected options: Future.DeriveOptions<T, To>,
+  ) {
+    this.underlying = underlying;
+    this.transformation = autorun('transformation', () => this.transform());
```
Comment (benbayard) [inline src/failable/derived.ts]: wow, you got the two way binding.

File: src/failable/derived.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  @computed
+  get isSuccess(): boolean {
+    return this.state === State.success;
+  }
+
+  /**
+   * Indicates if this Failable is a failure.
+   */
+  @computed
+  get isFailure(): boolean {
```
Comment (benbayard) [inline src/failable/derived.ts]: can this be `this is Failure`?

File: src/failable/derived.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  ): void {
+    this.state = state;
+    this.data = data;
+  }
+
+  /**
+   * Indicates if this Failable is a success.
+   */
+  @computed
+  get isSuccess(): boolean {
```
Comment (benbayard) [inline src/failable/derived.ts]: this is `Success<T>`?

File: src/failable/derived.ts (hunk 1 above)
Comment (kourge) [reply src/failable/derived.ts]: This is not truly two-way binding, but funny story: this doesn't work using `reaction`! I guess `reaction` is a lot more selective about when to notify observers, whereas `autorun` isn't.

Now that I think about it, `autorun`'s behavior is supposed to immediately run the function given in order to detect what values that function reads, so it should be safe to remove the `this.transform()` call below.

File: src/failable/derived.ts (hunk 3 above)
Comment (kourge) [reply src/failable/derived.ts]: The entire `Future<T>` interface specifies these as properties, so they can't be type guards, since only functions can be type guards :(

This is a separate discussion, but I'm open to adding type guards, but am stumped at what they'd be named.

**Comment by kourge** [comment]
@benbayard While writing tests for `Failable.prototype.rescue`, I realized that you cannot rescue a Failable into another type. In other words:

```ts
const f = new Failable<number>();
// This is possible:
const g = f.rescue(error => -1);
// This is not possible:
const h = f.rescue(error => error.message);
```

It's a pure type-level problem. In other words, any user of this API without a type system can already rescue a Failable into another type. Should we add this?

**Comment by kourge** [comment]
Mmm, that's a great solution. Let users rescue the failable into any they wish, but default to the same type as the one they started from.

**Comment by benbayard** [comment]
The new changes look solid to me. :shipit:

**Comment by kourge** [comment]
Hello! It's been a while. I've got some new commits. Here are the ones I said I would do:

- 137c8b7, Extend derivation support to `Loadable` as well. This involves restructuring the file system organization of it, just like `Failable`.
- 315da0c, Add the three new methods into the interfaces `Future` and `ReadonlyFuture`.

Along the way, a few more design decisions were made, which I will point out here:
- 9ccd825f changed `Loadable.MatchOptions` to be less vague. The previous intent was that a function that didn't take the second `loading` argument should be compatible, but the outcome made the second argument's type effectively become `boolean | undefined`. This was incorrect and forced functions to spell `!!loading` to use it properly.
- f0fd6b1f introduced type-level checks and refined them in 22126786. This is a way to make sure some type `T` is assignable to another type `U` without changing their definitions, but still get compiler errors. (We have co-evolving types, like `Future` and `ReadonlyFuture`, that are checked this way. We avoid using `Pick` to define these types because mapped types like `Pick` incur JSDoc loss.) These type-level checks are very reminiscent of a pattern in Go: to ensure that struct `S` is assignable to type `T`, one would write on the module level `var _ T = S{}`.
- ac8be0b4 added a `TraitSet` type. This is a type that represents the exact same data as `Loadable.State` does, but with a bitmap. (See the commit itself as to _why_ this is needed.) This commit added a lot of snapshot tests, which I've reviewed as a human, but I could use more eyes on them.

General things I'd love more eyes on:

- Grammatical errors and typos in documentation blocks. I found so much of these from existing docs while working on this!
- Tests. Areas to focus on are completeness (are we covering all bases?) and correctness (are the tests asserting the right things?).
- Design decisions. Am I accidentally making something way to complex? For example, for `TraitSet`, I struggled multiple times to avoid representing `Loadable.State` in other ways, but in the end, to preserve structural compatibility with `Failable`, I went ahead with it.

File: src/failable/derived.test.ts [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        },
+      });
+
+      expect(d.data).toBe(e);
+      expect(d.state).toBe(State.failure);
+    });
+  });
+
+  describe('given a failure-only transform', () => {
+    const options = {
```
Comment (benbayard) [inline src/failable/derived.test.ts]: Personally, I like `Object.freeze`ing immutable objects in tests, to ensure none of the code we have is mutating the options object (for instance)

Comment (kourge) [reply src/failable/derived.test.ts]: Good call.

File: src/failable/derived.test.ts [hunk 5]
```diff
@@ -0,0 +1,273 @@
+import {computed, useStrict, when} from 'mobx';
+
+import {Failable} from '.';
```
Comment (jbraithwaite) [inline src/failable/derived.test.ts]: Order inputs?

File: src/failable/match.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+export function match<T, A, B, C>(
+  state: Future.State,
+  data: T | Error | undefined,
+  options: Future.MatchOptions<T, A, B, C>,
+): A | B | C {
+  const {success, failure, pending} = options;
+
+  switch (state) {
+    case Future.State.success:
+      return success(data as T);
```
Comment (benbayard) [inline src/failable/match.ts]: why is this not automatically narrowed?

File: src/failable/derived.test.ts (hunk 5 above)
Comment (benbayard) [reply src/failable/derived.test.ts]: JUST LIKE OLD TIMES

File: src/failable/match.ts (hunk 6 above)
Comment (kourge) [reply src/failable/match.ts]: That sort of narrowing requires two pieces of data to live in the same object. So for example, if I wrote:

```ts
type FutureKernel<T> = 
  | {state: Future.State.success, data: T}
  | {state: Future.State.failure, data: Error}
  | {state: Future.State.fending, data: undefined};

const kernel: FutureKernel<T> = {state, data};
switch (kernel.state) {
  // ...
}
```

Then narrowing would happen automatically. Here I want to avoid incurring the overhead of creating a brand new object for the sole purpose of getting the right type. I could cast, which is what happened, or I could write a series of type guards:

```ts
function isSuccess<T>(state: Future.State, data: T | Error | undefined): data is T {
  return state === Future.State.success;
}
```

At this point, I'm comfortable with slightly murky type safety internally, as long as externally the type safety is sound.

File: src/failable/derived.test.ts (hunk 5 above)
Comment (kourge) [reply src/failable/derived.test.ts]: These imports are ordered, according to `tslint`. It's puzzling to me as to what order, though, because it doesn't look alphabetized. Just as a quick check I swapped these line orders and immediately got a lint warning.

File: src/failable/match.ts (hunk 6 above)
Comment (benbayard) [reply src/failable/match.ts]: okay. Thank you for the explanation. Maybe we could leave this as a comment on this code?

File: src/failable/derived.test.ts (hunk 5 above)
Comment (benbayard) [reply src/failable/derived.test.ts]: I think the new lines prohibit input ordering.

**Comment by benbayard** [comment]
This code is amazing @kourge . I looked for grammar mistakes, and none really stood out at me. The code itself looks great.

File: src/failable/derived.test.ts (hunk 5 above)
Comment (kourge) [reply src/failable/derived.test.ts]: https://github.com/palantir/tslint/blob/master/src/rules/orderedImportsRule.ts#L35

Comment (benbayard) [reply src/failable/derived.test.ts]: it seems like the last bit is what needs enforcing:

 Groups of imports are delineated by blank lines. You can use these to group imports
                however you like, e.g. by first- vs. third-party or thematically or can you can
                enforce a grouping of third-party, parent directories and the current directory.

I don't see the config for that enforcement though.

Comment (benbayard) [reply src/failable/derived.test.ts]:
```
Possible values for \`"grouped-imports"\` are:
            * \`false\`: Do not enforce grouping. (This is the default.)
            * \`true\`: Group source imports by \`"bar"\`, \`"../baz"\`, \`"./foo"\`.
```

Comment (benbayard) [reply src/failable/derived.test.ts]: So we would need to change:

https://github.com/UrbanDoor/mobx-failable/blob/master/tslint.json

to have:

```
"orderedInputs": [true, {"grouped-imports": true}]
```

and it should group every thing for us?

**Comment by benbayard** [comment]
🎆 🍾 🎆
